crates/oxc_linter/src/rules/valid_typeof.rs

```diff
@@ -0,0 +1,221 @@
+use oxc_ast::{
+    ast::{BinaryExpression, BinaryOperator, Expression, UnaryExpression, UnaryOperator},
+    AstKind, GetSpan, Span,
+};
+use oxc_diagnostics::{
+    miette::{self, Diagnostic},
+    thiserror::Error,
+};
+use oxc_macros::declare_oxc_lint;
+
+use crate::{context::LintContext, fixer::Fix, rule::Rule, AstNode};
+
+#[derive(Debug, Error, Diagnostic)]
+enum ValidTypeofDiagnostic {
+    #[error("eslint(valid-typeof): Typeof comparisons should be to string literals.")]
+    #[diagnostic(severity(warning))]
+    NotString(#[help] Option<&'static str>, #[label] Span),
+    #[error("eslint(valid-typeof): Invalid typeof comparison value.")]
+    #[diagnostic(severity(warning))]
+    InvalidValue(#[help] Option<&'static str>, #[label] Span),
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct ValidTypeof {
+    /// true requires typeof expressions to only be compared to string literals or other typeof expressions, and disallows comparisons to any other value.
+    require_string_literals: bool,
+}
+declare_oxc_lint!(
+    /// ### What it does
+    /// Enforce comparing `typeof` expressions against valid strings
+    ///
+    /// ### Why is this bad?
+    /// It is usually a typing mistake to compare the result of a typeof operator to other string literals.
+    /// ### Example
+    /// ```javascript
+    /// requireStringLiterals: false
+    /// incorrect:
+    /// typeof foo === "strnig"
+    /// correct:
+    /// typeof foo === "string"
+    /// typeof foo === baz
+    ///
+    /// requireStringLiterals: true
+    /// incorrect:
+    /// typeof foo === baz
+    /// ```
+    ValidTypeof,
+    nursery,
+);
+
+fn is_typeof_expr(expr: &Expression) -> bool {
+    if let Expression::UnaryExpression(unary) = expr && matches!(**unary, UnaryExpression { operator: UnaryOperator::Typeof, .. }){
+        true
+    } else {
+        false
+    }
+}
+
+impl Rule for ValidTypeof {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        if let AstKind::UnaryExpression(
+            unary @ UnaryExpression { operator: UnaryOperator::Typeof, .. },
+        ) = node.get().kind()
+        {
+            if let AstKind::BinaryExpression(BinaryExpression {
+                span: _,
+                left,
+                operator:
+                    BinaryOperator::Equality
+                    | BinaryOperator::Inequality
+                    | BinaryOperator::StrictEquality
+                    | BinaryOperator::StrictInequality,
+                right,
+            }) = ctx.parent_kind(node)
+            {
+                let valid_type = [
+                    "symbol",
+                    "undefined",
+                    "object",
+                    "boolean",
+                    "number",
+                    "string",
+                    "function",
+                    "bigint",
+                ];
```

Comment:
You can use a `phf` set here, see https://github.com/Boshen/oxc/blob/main/crates/oxc_linter/src/globals.rs